jerry-core/ecma/operations/ecma-promise-object.c

```diff
@@ -954,6 +954,269 @@ ecma_promise_then (ecma_value_t promise, /**< the promise which call 'then' */
+
+/**
+ * Definition of thrower function
+ *
+ * See also:
+ *         ES2020 25.6.5.3.2 step 8.
+ *
+ * @return ecma value
+ */
+ecma_value_t
+ecma_value_thunk_thrower_cb (const ecma_value_t function_obj, /**< the function itself */
+                             const ecma_value_t this_val, /**< this_arg of the function */
+                             const ecma_value_t args_p[], /**< argument list */
+                             const ecma_length_t args_count) /**< argument number */
+{
+  JERRY_UNUSED_3 (this_val, args_p, args_count);
+
+  ecma_object_t *func_obj_p = ecma_get_object_from_value (function_obj);
+  ecma_promise_value_thunk_t *value_thunk_obj_p = (ecma_promise_value_thunk_t *) func_obj_p;
+
+  jcontext_raise_exception (ecma_copy_value (value_thunk_obj_p->value));
+
+  return ECMA_VALUE_ERROR;
+} /* ecma_value_thunk_thrower_cb */
+
+/**
+ * Helper function for Then Finally and Catch Finally common parts
+ *
+ * See also:
+ *         ES2020 25.6.5.3.1 steps 2-7
+ *         ES2020 25.6.5.3.2 steps 2-7
+ *
+ * @return ecma value
+ */
+static ecma_value_t
+ecma_promise_than_catch_finally_helper (ecma_value_t function_obj) /**< the function itself */
+{
+  /* 2. */
+  ecma_object_t *func_obj_p = ecma_get_object_from_value (function_obj);
+  ecma_promise_finally_function_t *finally_func_obj = (ecma_promise_finally_function_t *) func_obj_p;
+
+  /* 3. */
+  JERRY_ASSERT (ecma_op_is_callable (finally_func_obj->on_finally));
+
+  /* 4. */
+  ecma_value_t result = ecma_op_function_call (ecma_get_object_from_value (finally_func_obj->on_finally),
+                                               ECMA_VALUE_UNDEFINED,
+                                               NULL,
+                                               0);
+
+  if (ECMA_IS_VALUE_ERROR (result))
+  {
+    return result;
+  }
+
+  /* 6. */
+  JERRY_ASSERT (ecma_is_constructor (finally_func_obj->constructor));
+
+  /* 7. */
+  ecma_value_t promise = ecma_promise_reject_or_resolve (finally_func_obj->constructor, result, true);
+
+  ecma_free_value (result);
+
+  return promise;
+} /* ecma_promise_than_catch_finally_helper */
+
+/**
+ * Definition of Then Finally Function
+ *
+ * See also:
+ *         ES2020 25.6.5.3.1
+ *
+ * @return ecma value
+ */
+ecma_value_t
+ecma_promise_then_finally_cb (const ecma_value_t function_obj, /**< the function itself */
+                              const ecma_value_t this_val, /**< this_arg of the function */
+                              const ecma_value_t args_p[], /**< argument list */
+                              const ecma_length_t args_count) /**< argument number */
+{
+  JERRY_UNUSED_3 (this_val, args_p, args_count);
+
+  /* 7. */
+  ecma_value_t promise = ecma_promise_than_catch_finally_helper (function_obj);
+
+  if (ECMA_IS_VALUE_ERROR (promise))
+  {
+    return promise;
+  }
+
+  /* 8. */
+  ecma_object_t *thunk_obj_p;
+  thunk_obj_p = ecma_create_object (ecma_builtin_get (ECMA_BUILTIN_ID_FUNCTION_PROTOTYPE),
```

Comment:
~~AFAIK to create the function object we could use the `ecma_op_create_external_function_object` method to avoid "manually" creating the function object.~~ Nevermind, the `sizeof` arg is different in that method.